linkerd/buffer/src/dispatch.rs

```diff
@@ -0,0 +1,136 @@
+    S::Error: Into<Error>,
+    S::Response: Send + 'static,
+    S::Future: Send + 'static,
+{
+    pub(crate) fn new(
+        inner: S,
+        rx: mpsc::Receiver<InFlight<Req, S::Future>>,
+        ready: watch::Sender<Poll<(), ServiceError>>,
+    ) -> Self {
+        Self {
+            inner: Some(inner),
+            rx,
+            ready,
+        }
+    }
+}
+
+macro_rules! return_ready {
+    () => {{
+        trace!("Complete");
+        return Ok(Async::Ready(()));
+    }};
+}
+
+macro_rules! return_ready_if {
+    ($cond:expr) => {{
+        if $cond {
+            return_ready!();
+        }
+    }};
+}
+
+impl<S, Req> Future for Dispatch<S, Req, S::Future>
+where
+    S: tower::Service<Req>,
+    S::Error: Into<Error>,
+    S::Response: Send + 'static,
+    S::Future: Send + 'static,
+{
+    type Item = ();
+    type Error = Never;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        // Complete the task when all services have dropped.
+        return_ready_if!(self.ready.poll_close().expect("must not fail").is_ready());
+
+        // Drive requests from the queue to the inner service.
+        loop {
+            let ready = match self.inner.as_mut() {
+                Some(inner) => inner.poll_ready(),
+                None => {
+                    // This is safe because ready.poll_close has returned NotReady.
+                    return Ok(Async::NotReady);
+                }
+            };
+
+            match ready {
+                // If it's not ready, wait for it..
+                Ok(Async::NotReady) => {
+                    return_ready_if!(self.ready.broadcast(Ok(Async::NotReady)).is_err());
+
+                    trace!("Waiting for inner service");
+                    return Ok(Async::NotReady);
+                }
+
+                // If the service fails, propagate the failure to all pending
+                // requests and then complete.
+                Err(error) => {
+                    let shared = ServiceError(Arc::new(error.into()));
+                    trace!(%shared, "Inner service failed");
+
+                    // First, notify services of the readiness change to prevent new requests from
+                    // being buffered.
+                    let is_active = self.ready.broadcast(Err(shared.clone())).is_ok();
+
+                    // Propagate the error to all in-
+                    while let Ok(Async::Ready(Some(InFlight { tx, .. }))) = self.rx.poll() {
+                        let _ = tx.send(Err(shared.clone().into()));
+                    }
+
+                    // Drop the inner Service to free its resources. It won't be used again.
+                    self.inner = None;
+
+                    // Ensure the task remains active until all services have observed the error.
+                    return_ready_if!(!is_active);
+
+                    // This is safe because ready.poll_close has returned NotReady.
+                    return Ok(Async::NotReady);
```

Comment:
Once a service has errored will it enter a non-error state? I would expect this to return `Ok(Async::Ready)`;

Reply:
On further thought this may make sense since as the comment mentions, `ready.poll_close()` returned `NotReady` above so after broadcasting an error to receivers this will be polled again once everything closes.

Reply:
That's right!